File: base-class-module/src/main/java/jingya/com/base_class_module/BaseMvps/BaseMvpLazyLoadingFragment.java
package jingya.com.base_class_module.BaseMvps;

import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.greenrobot.eventbus.EventBus;

/**
 * @author kuky
 * @description MVP 下懒加载 fragment 基类
 */
public abstract class BaseMvpLazyLoadingFragment<V extends BaseMvpViewImpl, P extends BaseMvpPresenter<V>,
        VB extends ViewDataBinding> extends Fragment {

    protected VB mViewBinding;
    protected P mPresenter;
    protected boolean isPageVisible;
    protected boolean isPageCreated;

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        isPageCreated = true;
        tryLoad();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (enabledEventBus())
            EventBus.getDefault().register(this);

        mViewBinding = DataBindingUtil.inflate(inflater, getLayoutId(), container, false);
        initFragment(savedInstanceState);
        mPresenter = initPresenter();
        mPresenter.attach((V) this);
        presenterActions();
        setListener();
        return mViewBinding.getRoot();
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (isVisibleToUser) {
            isPageVisible = true;
            tryLoad();
        } else {
            isPageVisible = false;
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        mPresenter.onResume();
    }

    @Override
    public void onStop() {
        super.onStop();
        mPresenter.onStop();
    }

    @Override
    public void onPause() {
        super.onPause();
        mPresenter.onPause();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (enabledEventBus())
            EventBus.getDefault().unregister(this);
        mPresenter.detach();
    }

    private void tryLoad() {
        if (isPageVisible && isPageCreated) {
            lazyLoading();
            isPageVisible = false;
            isPageCreated = false;
        }
    }

    protected abstract P initPresenter();

    protected abstract int getLayoutId();

    protected abstract void initFragment(Bundle savedInstanceState);

    protected abstract void lazyLoading();

    protected boolean enabledEventBus() {
        return false;
    }

    protected void presenterActions() {

    }

    protected void setListener() {

    }
}

File: base-kt-module/src/main/java/com/kuky/base_kt_module/BaseMvps/BaseMvpLazyLoadingFragment.kt
package com.kuky.baselib.baseMvpClass

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

/**
 * @author kuky
 * @description
 */
abstract class BaseMvpLazyLoadingFragment<in V : BaseMvpViewImpl, P : BaseMvpPresenter<V>, VB : ViewDataBinding> : Fragment() {
    protected lateinit var mViewBinding: VB
    protected lateinit var mPresenter: P
    protected var isPageCreated: Boolean = false
    protected var isPageVisible: Boolean = false

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        isPageCreated = true
        tryLoad()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mViewBinding = DataBindingUtil.inflate(inflater, getLayoutId(), container, false)
        initFragment(savedInstanceState)
        mPresenter = initPresenter()
        mPresenter.attachView(this as V)
        presenterActions()
        setListener()
        handleRxBus()
        return mViewBinding.root
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (isVisibleToUser) {
            isPageVisible = true
            tryLoad()
        } else {
            isPageVisible = false
        }
    }

    override fun onResume() {
        super.onResume()
        mPresenter.onResume()
    }

    override fun onStop() {
        super.onStop()
        mPresenter.onStop()
    }

    override fun onPause() {
        super.onPause()
        mPresenter.onPause()
    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter.detachView()
    }

    private fun tryLoad() {
        if (isPageCreated && isPageVisible) {
            lazyLoading()
            isPageCreated = false
            isPageVisible = false
        }
    }

    protected abstract fun getLayoutId(): Int

    protected abstract fun initFragment(savedInstanceState: Bundle?)

    protected abstract fun initPresenter(): P

    protected abstract fun lazyLoading()

    protected open fun presenterActions() {}

    protected open fun setListener() {}

    protected open fun handleRxBus() {}
}
